### PR TITLE
FEATURE: SD-870 viewport configuration

### PR DESCRIPTION
**Pull request by ptquang86**
---

[Check PR owner responsibilities](https://confluence.intgdc.com/display/Development/Code-reviews#Code-reviews-Ownerresponsibilities)

Supported PR commands:

Command | Description
--- | ---
`extended test - examples` | Live examples tests
`extended test - storybook` | Storybook screenshot tests

See more [options](https://confluence.intgdc.com/display/kbhr/How+to+work+with+git+and+Github#HowtoworkwithgitandGithub-Extendedchecks).

# PR Checklist
- [x] Verify code changes ([Checklist](https://confluence.intgdc.com/display/Development/Code-reviews+checklist), [Best practices](https://confluence.intgdc.com/display/Development/Code-reviews+best+practices))
- [x] [Verify pull-request formalities](https://confluence.intgdc.com/display/Development/Code-reviews)
- [x] Change was tested by using [gdc-dev-release](https://confluence.intgdc.com/display/~tomas.vojtasek/Private+NPM) in [gdc-analytical-designer](https://github.com/gooddata/gdc-analytical-designer) and [gdc-dashboards](https://github.com/gooddata/gdc-dashboards) (if applicable)
- [x] Migration guide (for major changes) is mentioned in [CHANGELOG.md](../blob/master/CHANGELOG.md).
- [x] Successful `extended test - examples`
- [x] Successful `extended test - storybook`
- [x] Checked yarn.lock consistency (no dep. duplicities especially Goodstrap)
- [x] Squash commits

# Related PRs
- gdc-analytical-designer: https://github.com/gooddata/gdc-analytical-designer/pull/2882
- gdc-dashboards: https://github.com/gooddata/gdc-dashboards/pull/2615

# Related Jira tasks
- SD-875: https://jira.intgdc.com/browse/SD-875

### CODE REVIEW

**Comment by ptquang86** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/6507/ : SUCCESS in 5m 53s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/2306/ : SUCCESS in 4s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3377/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/2307/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2780/ : UNSTABLE in 18m 01s

**Comment by ptquang86** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/6508/ : SUCCESS in 5m 27s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/2308/ : SUCCESS in 4s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3378/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/2309/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2781/ : SUCCESS in 15m 25s

**Comment by ptquang86** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-1446.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/2310/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2782/ : SUCCESS in 28m 52s

File: stories/core_components/GeoPushpinChart.tsx [hunk 1]
```diff
@@ -68,6 +68,20 @@ storiesOf("Core components/GeoPushpinChart", module)
             }),
         ),
     )
+    .add("with location and viewport", () =>
```
Comment (naibinh) [inline stories/core_components/GeoPushpinChart.tsx]: `with location and viewport` -> `with North America viewport`

File: src/components/core/geoChart/GeoChartRenderer.tsx [hunk 2]
```diff
@@ -113,29 +113,15 @@ export default class GeoChartRenderer extends React.Component<IGeoChartRendererP
...
         this.chart = new mapboxgl.Map({
             ...DEFAULT_MAPBOX_OPTIONS,
-            bounds,
             container: this.chartRef,
-            center: center || DEFAULT_CENTER,
             // If true, the map’s canvas can be exported to a PNG using map.getCanvas().toDataURL().
             // This is false by default as a performance optimization.
             preserveDrawingBuffer: isExportMode,
-            zoom,
+            ...getViewportOptions(data, config),
```
Comment (naibinh) [inline src/components/core/geoChart/GeoChartRenderer.tsx]: I think spread operator should come first, just right after `... DEFAULT_MAPBOX_OPTIONS ` is fine :)

File: src/constants/geoChart.ts [hunk 3]
```diff
@@ -3,6 +3,16 @@ import mapboxgl from "mapbox-gl";
...
 
+export const areaBounds: { [key: string]: mapboxgl.LngLatBoundsLike } = {
+    africa: [{ lat: -36.3318, lng: -20.2255 }, { lat: 37.982, lng: 54.4841 }],
+    asia: [{ lat: -9.8256, lng: 24.3532 }, { lat: 79.488044, lng: 180 }],
+    australia: [{ lat: -47.9436, lng: 109.8601 }, { lat: 0, lng: 180 }],
+    europe: [{ lat: 33.6033, lng: -26.5876 }, { lat: 77.442, lng: 78.3182 }],
+    north_america: [{ lat: 2.2236, lng: -170.4181 }, { lat: 72.0157, lng: -52.69208 }],
+    south_america: [{ lat: -56.8929, lng: -90.242 }, { lat: 14.0995, lng: -31.3784 }],
+    world: [{ lat: -84, lng: -180 }, { lat: 84, lng: 180 }],
+};
```
Comment (naibinh) [inline src/constants/geoChart.ts]: Just my opinion, `areaBounds ` -> `VIEWPORTS` as they are all constant

File: stories/core_components/GeoPushpinChart.tsx [hunk 4]
```diff
@@ -68,6 +68,20 @@ storiesOf("Core components/GeoPushpinChart", module)
...
+                location: ATTRIBUTE_LOCATION_GEOCHART,
+                config: {
+                    ...DEFAULT_CONFIG,
+                    viewport: {
+                        area: "north_america",
+                    },
+                },
+            }),
+        ),
+    )
```
Comment (naibinh) [inline stories/core_components/GeoPushpinChart.tsx]: We should have two more cases with same above data
1. with World viewport
2. with 'Include all data' viewport

**Comment by ptquang86** [comment]
extended test - storybook

File: stories/core_components/GeoPushpinChart.tsx (hunk 1 above)
Comment (ptquang86) [reply stories/core_components/GeoPushpinChart.tsx]: updated

File: src/components/core/geoChart/GeoChartRenderer.tsx (hunk 2 above)
Comment (ptquang86) [reply src/components/core/geoChart/GeoChartRenderer.tsx]: updated

File: src/constants/geoChart.ts (hunk 3 above)
Comment (ptquang86) [reply src/constants/geoChart.ts]: updated. Thanks

File: stories/core_components/GeoPushpinChart.tsx (hunk 4 above)
Comment (ptquang86) [reply stories/core_components/GeoPushpinChart.tsx]: New stories has been added. Thanks :)

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/6512/ : SUCCESS in 12m 44s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/2315/ : SUCCESS in 5s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3380/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/2316/ : SUCCESS in 34s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2784/ : UNSTABLE in 24m 09s

**Comment by ptquang86** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/6517/ : SUCCESS in 11m 24s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/2321/ : SUCCESS in 6s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3382/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/2322/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2786/ : SUCCESS in 21m 32s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/6527/ : SUCCESS in 7m 35s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/2330/ : SUCCESS in 5s

File: src/constants/geoChart.ts [hunk 5]
```diff
@@ -3,6 +3,16 @@ import mapboxgl from "mapbox-gl";
 import { DEFAULT_COLORS } from "../components/visualizations/utils/color";
 import { IGeoLngLat } from "../interfaces/GeoChart";
 
+export const VIEWPORTS: { [key: string]: mapboxgl.LngLatBoundsLike } = {
```
Comment (lvkhanh) [inline src/constants/geoChart.ts]: should seperate `{ [key: string]: mapboxgl.LngLatBoundsLike } `to type, such as `IGeoViewportConfig`

File: src/helpers/geoChart/viewport.ts [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+interface IGeoViewport {
+    bounds?: mapboxgl.LngLatBoundsLike;
+    center?: IGeoLngLat;
+    zoom?: number;
+}
+
+export function getViewportOptions(data: IGeoLngLat[], config: IGeoConfig): IGeoViewport {
+    const center: IGeoLngLat = get<IGeoConfig, "center">(config, "center");
+    const zoom: number = get<IGeoConfig, "zoom", number>(config, "zoom", DEFAULT_ZOOM);
+    const { area }: IGeoConfigViewport = get<IGeoConfig, "viewport", {}>(config, "viewport", {});
```
Comment (lvkhanh) [inline src/helpers/geoChart/viewport.ts]: Default value is empty object, is it correct? I see we define type of `VIEWPORTS` is `{ [key: string]: mapboxgl.LngLatBoundsLike }`, so VIEWPORTS[{}] still `return undefined`

File: src/interfaces/GeoChart.ts [hunk 7]
```diff
@@ -54,6 +54,20 @@ export interface IGeoData {
...
 
+export type IGeoConfigViewportArea =
+    | "auto" // default, Include all data
+    | "continent_af"
+    | "continent_as"
+    | "continent_au"
+    | "continent_eu"
+    | "continent_na"
+    | "continent_sa"
+    | "world";
```
Comment (naibinh) [inline src/interfaces/GeoChart.ts]: Please give comment for these codes like [here](https://github.com/gooddata/gooddata-react-components/pull/1446/files#diff-fa62e9a6886bb2448106a41bbcfa6e89R7)

**Comment by naibinh** [comment]
@ptquang86 , please also add example item and test for it :)

File: src/constants/geoChart.ts (hunk 5 above)
Comment (ptquang86) [reply src/constants/geoChart.ts]: We already have `IGeoConfigViewport`, so I use `IGeoViewports` instead :)

Comment (ptquang86) [reply src/constants/geoChart.ts]: updated. thanks :)

File: src/helpers/tests/geoChart/viewport.spec.ts [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+                    area: "world",
+                },
+            };
+            expect(getViewportOptions(data, config)).toEqual({
+                center,
+                zoom: 2, // default zoom
+            });
+        });
+
+        it("should return bounds for passed area", () => {
```
[the remark this answers is not in the document]
Comment (ptquang86) [reply src/helpers/tests/geoChart/viewport.spec.ts]: updated. thanks :)

File: src/interfaces/GeoChart.ts (hunk 7 above)
Comment (ptquang86) [reply src/interfaces/GeoChart.ts]: updated.

**Comment by ptquang86** [comment]
Hi @naibinh @lvkhanh I have updated the code + added examples.
Please help me review again.
Thanks

**Comment by ptquang86** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-1446.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check-extended aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by ptquang86** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/6536/ : SUCCESS in 8m 47s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/2339/ : SUCCESS in 20s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3390/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/2340/ : SUCCESS in 15s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2790/ : UNSTABLE in 18m 11s

**Comment by ptquang86** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/6537/ : SUCCESS in 16m 02s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/2341/ : SUCCESS in 7s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3391/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/2342/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2791/ : SUCCESS in 18m 14s

**Comment by ptquang86** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-1446.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by naibinh** [comment]
LGTM

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/2343/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2792/ : SUCCESS in 26m 08s

**Comment by lvkhanh** [comment]
LGTM

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3392/ : SUCCESS in 17m 52s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/6539/ : SUCCESS in 10m 26s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/1575/ : SUCCESS in 15s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/2266/ : SUCCESS in 23m 06s

**Pull request merged**